pkg/images/images.go

```diff
@@ -148,30 +152,51 @@ func getImages(distro string, versions []interface{}) (all []string, err error)
 				all = append(all, converted)
 			}
 		case utiliies.RKE2, utiliies.K3S:
-			temp, _ := version.(map[string]interface{})
-			v, _ := temp["version"].(string)
-			for _, os := range []string{linux, window} {
-				url, _ := urls[distro][os]
-				if url != "" {
-					images, err := getImagesFromURL(fmt.Sprintf(url, v))
-					if err != nil {
-						return nil, fmt.Errorf("failed to get images: %o", err)
-					}
-					if len(images) > 0 {
-						logrus.Tracef("distro %s version %s adds %s ", distro, version, images)
-						all = append(all, images...)
+			wg.Add(1)
+			go func(version interface{}) {
+				defer wg.Done()
+				temp, _ := version.(map[string]interface{})
+				v, _ := temp["version"].(string)
+
+				for _, os := range []string{linux, window} {
+					url := urls[distro][os]
+					if url != "" {
+						images, err := getImagesFromURL(fmt.Sprintf(url, v))
+						if err != nil {
+							errCh <- fmt.Errorf("failed to get images: %o", err)
+							return
+						}
+						if len(images) > 0 {
+							logrus.Tracef("distro %s version %s adds %s", distro, version, images)
+							mu.Lock()
+							all = append(all, images...)
+							mu.Unlock()
+						}
 					}
 				}
-			}
-			// add the following images which are not in the upstream images.txt file
-			// - rancher/rancher-<distro>-upgrade
-			// - rancher/system-agent-installer-<distro>
-			safeVersion := strings.ReplaceAll(v, "+", "-")
-			upgradeImage := fmt.Sprintf(upgradeImage, strings.ToLower(distro), safeVersion)
-			systemAgentInstallerImage := fmt.Sprintf(systemAgentInstallerImage, strings.ToLower(distro), safeVersion)
-			all = append(all, upgradeImage, systemAgentInstallerImage)
+				// add the following images which are not in the upstream images.txt file
+				// - rancher/rancher-<distro>-upgrade
+				// - rancher/system-agent-installer-<distro>
+				safeVersion := strings.ReplaceAll(v, "+", "-")
+				upgradeImage := fmt.Sprintf(upgradeImage, strings.ToLower(distro), safeVersion)
+				systemAgentInstallerImage := fmt.Sprintf(systemAgentInstallerImage, strings.ToLower(distro), safeVersion)
+				mu.Lock()
+				all = append(all, upgradeImage, systemAgentInstallerImage)
+				mu.Unlock()
+			}(version)
+		}
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	// Check for errors
+	for err := range errCh {
+		if err != nil {
+			return nil, err
```

Comment:
I suppose if we're erroring out on any one error, we don't need to wait for all versions to be fetched. User would need to run `go generate` in that case anyway.

Reply:
I have modified the code to use `context.WithCancel` to cancel all other goroutines once an error is encountered, additionally `errCh` is updated to size 1 to capture and immediately return the first error encountered.